Move project tree to an application-driven logic

jmaupetit opened the pull request:
## Purpose

We need to make applications easily pluggable to use Arnold as a generic way to deploy them. This means that we need to move the project tree to a structure easily overridable when using Arnold's docker image. Moreover, we need to be able to easily deploy a single application without having to comment or update group vars definitions.

## Proposal

- [x] move OpenShift object templates to the `apps/` directory
- [x] move ConfigMaps to the target application configmaps directory (_e.g._  `apps/{{ app }}/configmap`)
- [x] ~move secrets to the target application secrets directory (_e.g._  `apps/{{ app }}/secret`)~ edit: postponed to a future PR
- [x] ~move application-specific settings to a file (_e.g._  `apps/{{ app }}/group_vars/main.yml`)~ edit: postponed to a future PR
- [x] create application objects by scanning _ad hoc_ directories instead of hard-coding them in a configuration (_i.e._ `apps/{{ app }}/dc/*.yml.j2` for DCs) \*
- [x] ~get rid of the `_common` templates?~
- [x] ~update the documentation~ edit: postponed to a future PR

\* we need to find a way to easily modify this mechanism for a particular environment, _e.g._ ignore endpoints in development.

This work is related to the following issues: #40, #44

jmaupetit committed:
🚚(templates) initiate an app-driven tree

This is the first step of an app-driven logic: all templates for an
application should live in a "apps/{{ app }}/" directory. All OpenShift
objects specific to an app will live in the apps/{{ app }}/{{ object }}
directory.

jmaupetit committed:
✨(configmap) make configurations app-specific

* Move application configmaps to the ad hoc directory, e.g.
  apps/richie/configmap/nginx/richie.conf.j2
* Set the applications list as an "apps" global variable that can be
  overriden via extra vars from command line (woot
* Iterate over applications to create all configmaps during deployment
* Fix configmap names in DCs or jobs to fit with the following pattern:
  {{ app }}-{{ service }}-{{ deployment_stamp }}

jmaupetit committed:
🚚(templates) restore a per-app templates directory

Explicitly move OpenShift's object templates to an app/templates
directory so that Ansible users won't be disturbed by applications tree.

jmaupetit committed:
✨(apps) make applications to deploy discoverable

* Make configMaps service-specific
* Make volumes (pvc) application-specific
* Add all required services templates for apps so that they are
  self-contained and do not depend on a "common" template
* Move playbooks tasks to a separated tasks sets located in tasks; those
  sets have been designed with a single app responsibility. Playbooks
  must iterate over applications with this tasks sets.
* Add a lookup_apps plugin to handle applications discoverability and
  generate the "apps" variable that describe all available apps to
  deploy.

jmaupetit commented:
The current `apps` directory tree looks like:

```
apps
├── _common
│   └── templates
│       ├── rolebinding
│       │   └── add-role-to-sa.yml.j2
│       ├── route
│       │   └── main.yml.j2
│       └── secret
│           └── env_vars.yml.j2
├── edxapp
│   └── templates
│       ├── cms
│       │   ├── _configs
│       │   │   ├── create_demo_users.py.j2
│       │   │   ├── docker_run.py.j2
│       │   │   └── settings.yml.j2
│       │   ├── dc.yml.j2
│       │   ├── job_collectstatic.yml.j2
│       │   ├── job_db_migrate.yml.j2
│       │   ├── job_load_fixtures.yml.j2
│       │   ├── route.yml.j2
│       │   └── svc.yml.j2
│       ├── lms
│       │   ├── _configs
│       │   │   ├── docker_run.py.j2
│       │   │   └── settings.yml.j2
│       │   ├── dc.yml.j2
│       │   ├── job_collectstatic.yml.j2
│       │   ├── job_db_migrate.yml.j2
│       │   ├── route.yml.j2
│       │   └── svc.yml.j2
│       ├── memcached
│       │   ├── dc.yml.j2
│       │   └── svc.yml.j2
│       ├── mongodb
│       │   ├── dc.yml.j2
│       │   ├── ep.yml.j2
│       │   └── svc.yml.j2
│       ├── mysql
│       │   ├── dc.yml.j2
│       │   ├── ep.yml.j2
│       │   └── svc.yml.j2
│       ├── nginx
│       │   ├── _configs
│       │   │   ├── cms.conf.j2
│       │   │   └── lms.conf.j2
│       │   ├── dc.yml.j2
│       │   └── svc.yml.j2
│       └── _volumes
│           ├── data.yml.j2
│           ├── media.yml.j2
│           └── static.yml.j2
├── redirect
│   └── templates
│       └── nginx
│           ├── _configs
│           │   └── aliases_routes.conf.j2
│           ├── dc.yml.j2
│           ├── route.yml.j2
│           └── svc.yml.j2
└── richie
    ├── templates
    │   ├── elasticsearch
    │   │   ├── dc.yml.j2
    │   │   └── svc.yml.j2
    │   ├── nginx
    │   │   ├── _configs
    │   │   │   └── richie.conf.j2
    │   │   ├── dc.yml.j2
    │   │   └── svc.yml.j2
    │   ├── postgresql
    │   │   ├── dc.yml.j2
    │   │   ├── ep.yml.j2
    │   │   └── svc.yml.j2
    │   ├── richie
    │   │   ├── dc.yml.j2
    │   │   ├── job_collectstatic.yml.j2
    │   │   ├── job_db_migrate.yml.j2
    │   │   ├── job_regenerate_indexes.yml.j2
    │   │   ├── route.yml.j2
    │   │   └── svc.yml.j2
    │   └── _volumes
    │       ├── media.yml.j2
    │       └── static.yml.j2
    └── vars
        └── main.yml

30 directories, 55 files
```

sampaccoud commented on plugins/utils/lookup_apps.py:
```diff
@@ -0,0 +1,164 @@
+        │   │   ├── dc.yml.j2
+        │   │   ├── ep.yml.j2
+        │   │   └── svc.yml.j2
+        │   ├── richie
```
Could be called "app" to be generic?

jmaupetit replied:
I'm not sure since the service name can be "injected" in its templates.

sampaccoud replied:
You already have the name "richie" from the parent directory... can you get it from there?  It would be nice if the structure inside an app could be fixed.

jmaupetit replied:
The service name has an impact on related objects, hence, you are free to name a service `app` if you expect a configMap called `richie-app-{{ deployment_stamp }}`. It's up to the application's developer to be consistent in it's naming conventions.

That said, i think there is room for improvement to make our templates more generic.

cchaudier replied:
Richie is the name of the app but it's also the name of the main service in this app.
For the moment it's not a problem IMHO.
In the future I think whe can work on a app with only one main service (like an Ansible role) in the dir and a dependencies file (like the [meta/main.yml](https://docs.ansible.com/ansible/latest/reference_appendices/galaxy.html?highlight=meta#dependencies) in role dir).

jmaupetit commented:
This PR is already too big for a comfortable review. Sorry for that. I've postponed some tasks to other PRs. My first intent was to make apps "discoverable" in this work; this is now done at 80% (application secrets and vars are not implemented yet).

cchaudier commented on create_config.yml:
```diff
@@ -6,4 +6,8 @@
 
     - import_tasks: tasks/set_vars.yml
-    - import_tasks: tasks/create_config.yml
+    - include_tasks: tasks/create_app_config.yml
```
Why don't use tasks/run_tasks_for_apps.yml helper ?

jmaupetit replied:
You are right: it's not consistent. Let's remove this unused playbook instead.

cchaudier reviewed (commented):
Very great job !
I've juste one question.

cchaudier commented on plugins/utils/lookup_apps.py:
```diff
@@ -0,0 +1,161 @@
+                "name": "richie",
+                "services": [
+                    {
+                        "configs": [
```
In my head the template are not in this var but find by the playbooks with the app.name and service.name.
If we defined the apps var we don't need to liste the template on the var.
For me the data structure must be :
```
 [
            {
                "name": "richie",
                "services": [
                    {
                        "name": "nginx"
                    },
                    {
                        "name": "postgresql"
                    },
                    {
                        "name": "elasticsearch"
                    },
                    {
                        "name": "richie"
                    }
                ]
            }
        ]
```

jmaupetit replied:
We've discussed this. I'll make a new PR to improve apps configuration overriding mechanism.

cchaudier reviewed (commented):
In my head the apps var don't contain the templates.

jmaupetit committed:
🔥(playbooks) remove the create_config playbook

We are supposed to create configMaps while deloying an application, so
there is not need to keep this playbook.

cchaudier commented on docs/developer_guide/playbooks.md:
```diff
@@ -65,11 +65,11 @@ OpenShift objects for a customer (default: `patient0`) in a particular
-* `create_objects.yml`
+- `create_project.yml`
+- `create_secrets.yml`
+- `create_config.yml`
```
Juste remove this line before merge ;)

jmaupetit replied:
Good catch, thx!
